### DISCUSSION TITLE
A simple feature request (with bounty/reward)

### DISCUSSION THREAD

**Discussion by osa1** [neovim/neovim#13826] [Ideas]
It's very distracting when the viewport moves back and forth as I type the search term in `/`, and moves again when I finish typing it, with `<CR>`.

As far as I know there's no way to disable this without resorting to various hacks (like marking the current cursor location and jumping back automatically on `<CR>`) and none of the hacks work great.

I'd like a setting to disable this. The viewport should not move at all as I type the search term and after `<CR>`. My cursor should also not move at all (when typing the search term and after `<CR>`). `hlsearch`, `incsearch`, and other search related settings should work as before.

How this is enabled does not matter as long as it's somehow possible to do in `init.vim`.

I have a few questions about this feature request:

- Does this sound reasonable to neovim devs? Would neovim devs be willing to accept a patch that properly implements this?
- Is there a way to offer bounty/reward for the implementation of this? (to be paid when the patch is merged to master) Can I use [bountysource][1] for this?
- What would be a reasonable bounty for this? I have something like $50 in mind, but I'd be willing to go higher than that.

Thanks.

[1]: https://www.bountysource.com/teams/neovim/issues

**Comment by fsouza** [comment]
(Disclaimer: I am not a maintainer)

I'm confused about this. Are you looking for the ability to search only within the current viewport?

If you don't want your cursor to jump as you type, you can set noincsearch, but I don't quite understand what's the behavior you're looking for when you hit `<cr>`.

**Reply by osa1** [reply]
> Are you looking for the ability to search only within the current viewport?

No, it should search in all buffers as it currently does, but my viewport should not be changed as I type the search term or after `<CR>`.

> If you don't want your cursor to jump as you type, you can set noincsearch, but I don't quite understand what's the behavior you're looking for when you hit <cr>.

I want `incsearch` to work as before, I just want viewport to not move, as I type the search term and after `<CR>`. After `<CR>` the matched text should be highlighted as in (n)vim today.

Really everything should work the same, except my viewport should not move. 

I don't know how to explain this better..

**Reply by osa1** [reply]
Suppose I'm editing a file which has 1000 lines, and I'm currently seeing lines 50 to 100.

Now I start typing a search term `/bl` and there's a match in line 1. Now vim shows me lines 1-50 to show the matching line.

I don't want this to happen, I want to keep seeing lines 50-100 as I type the search term. It doesn't matter how many matches there are, or which lines the matches are on.

As I finish my search term with `/blah<CR>` I still want to be seeing lines 50-100.

However any matches in lines 50-100 (and others, in other buffers shown in splits etc.) should still be highlighted. Basically everything should work as before, just the viewport should not ever change.

Is this more clear?

**Comment by chrisbra** [comment]
> As I finish my search term with /blah<CR> I still want to be seeing lines 50-100.

Not sure what you mean with `finished`. After a successful search, Vim will always position the cursor on the match. That is what a search is for, isn't it? You can however disable that Vim starts positioning the cursor *as soon as you type your search term*, use `:set noincsearch` for that. And you can also force a match, something like `/foobar\&\%>49l\&\%<60l` this assert that the match is between line 50 and 59.

**Reply by osa1** [reply]
> Not sure what you mean with finished

I just mean when I press "enter" (`<CR>`) after typing the search term.

> After a successful search, Vim will always position the cursor on the match.

This is what I'm trying to avoid.

> That is what a search is for, isn't it?

I want to see occurrences of the searched term, and I want to be able to jump to next/prev match using `n`/`N` (or with `C-t`/`C-g` in `incsearch`, but the viewport should not move until I press `C-t`/`C-g`). You don't have to jump to a match for it to be useful. I'm sure 90% of the time first thing I do after searching something is `C-o` to jump back where I was (though that doesn't work perfectly as my viewport still moves and I get distracted).

> You can however disable that Vim starts positioning the cursor as soon as you type your search term, use :set noincsearch for that.

I want `incsearch` (and everything else) to work as before. I just want the viewport to stay still.

> And you can also force a match, something like /foobar\&\%>49l\&\%<60l this assert that the match is between line 50 and 59.

What do you mean by "assert"? Only search between those lines? Or something else?

**Reply by chrisbra** [reply]
> This is what I'm trying to avoid.

Why are you searching then? I guess you want to use the `search()` function then, if you want a more low-level functionality.

> What do you mean by "assert"? Only search between those lines? 

yes, only between those lines. Please read the help.

**Reply by osa1** [reply]
> Why are you searching then?

I want to matched terms to be highlighted, `C-n` and `C-N` to work etc. It's still useful. I emulate this "feature" manually all day every day. It's useful to me.

> yes, only between those lines. Please read the help.

Which help page/section?

I'll look at `search()`.

**Reply by chrisbra** [reply]
patterns/regular expression atoms are described at `:h pattern.txt`

It you want to highlight specific items, you can use `matchadd()`. Neovim probably has lua bindings for that as well, don't know.

Alternatively, you could set the search register like this `:let @/='foobar'` and jump to matches later using `n`, `N`. Not sure, why you mean with `<C-N>` and `<C-n>`. And you can always save and restore the viewport using `winsaveview()/winrestview()` calls.

**Comment by osa1** [comment]
If anyone else is looking for this feature, I just realized that vscode has a setting for this which is called `editor.find.cursorMoveOnType` in the JSON file. The setting in the settings UI is as below.

![Screenshot_2021-02-03_19-18-42](https://user-images.githubusercontent.com/448274/106775937-a49d7280-6654-11eb-9110-a077255ffd99.png)
